File: CppEIM/src/regression/Regression.cpp
#include <regression/Regression.h>

inline void normalize_fn(mat &features, vec &outputs, vec &f_mean, vec &f_std,
                         double &o_mean, double &o_std, uword bias_entry){
    f_mean = mean(features, 1);
    f_std = stddev(features, 0, 1);
    if (bias_entry >= 0){
        f_mean[bias_entry] = 0.0;
        f_std[bias_entry] = 1.0;
    }
    features = (features.each_col() - f_mean).each_col() / f_std;
    o_mean = mean(outputs);
    o_std = stddev(outputs);
    outputs = (outputs - o_mean) / o_std;
}

inline void undo_normalization(vec &params, const vec &f_mean, const vec &f_std, const double o_mean,
                               const double o_std, uword bias_entry){
    params %= (o_std / f_std);
    params[bias_entry] = params[bias_entry] - dot(f_mean, params) + o_mean;
}

inline vec regress(const mat &weighted_features, const mat &features, const vec &outputs, double reg_fact){
    mat reg_mat = reg_fact * mat(features.n_rows, features.n_rows, fill::eye);
    vec params;
    bool succ = solve(params, weighted_features * features.t() + reg_mat, weighted_features * outputs);
    if (succ){
        return params;
    } else {
        throw std::logic_error("Inverting Matrix during fitting linear model failed!");
    }
}

inline vec linear_regression(mat &features, vec &outputs, double reg_fact, bool normalize, uword bias_entry){
    if(normalize) {
        vec f_mean, f_std;
        double o_mean, o_std;
        normalize_fn(features, outputs, f_mean, f_std, o_mean, o_std, bias_entry);
        vec params = regress(features, features, outputs, reg_fact);
        undo_normalization(params, f_mean, f_std, o_mean, o_std, bias_entry);
        return params;
    } else {
        return regress(features, features, outputs, reg_fact);
    }
}

inline vec weighted_linear_regression(mat &features, vec &outputs, const rowvec& weights, double reg_fact, bool normalize, uword bias_entry){
    if (normalize){
        vec f_mean, f_std;
        double o_mean, o_std;
        normalize_fn(features, outputs, f_mean, f_std, o_mean, o_std, bias_entry);
        mat weighted_features = features.each_row() % weights;
        vec params = regress(weighted_features, features, outputs, reg_fact);
        undo_normalization(params, f_mean, f_std, o_mean, o_std, bias_entry);
        return params;
    } else {
        mat weighted_features = features.each_row() % weights;
        return regress(weighted_features, features, outputs, reg_fact);
    }
}

/*** BASE CLASS REGRESSION FUNCTION ***/

RegressionFunc::RegressionFunc(double reg_fact, bool normalize) :
reg_fact(reg_fact),
normalize(normalize) {}

mat RegressionFunc::call(const mat &input) const{
    if(std::isnan(params.at(0, 0))){
        throw std::logic_error("Model needs to be fitted first");
    }
    mat features = feature_fn(input);
    return (params.t() * feature_fn(input));
}

void RegressionFunc::fit(const mat &inputs, const vec &outputs) {
    mat features = feature_fn(inputs);
    vec outputs_copy = outputs;
    params = linear_regression(features, outputs_copy, reg_fact, normalize, bias_entry());
}

void RegressionFunc::fit(const mat &inputs, const mat &outputs) {
    mat features = feature_fn(inputs);
    params = mat(features.n_rows, outputs.n_rows);
    for(uword i = 0; i < outputs.n_rows; ++i){
        vec outputs_copy = outputs.row(i).t();
        params.col(i) = linear_regression(features, outputs_copy, reg_fact, normalize, bias_entry());
    }
}


void RegressionFunc::fit(const mat &inputs, const vec &outputs, const rowvec &weights) {
    mat features = feature_fn(inputs);
    vec outputs_copy = outputs;
    params = weighted_linear_regression(features, outputs_copy, weights, reg_fact, normalize, bias_entry());
}

void RegressionFunc::fit(const mat &inputs, const mat &outputs, const mat &weights) {
    mat features = feature_fn(inputs);
    params = mat(features.n_rows, outputs.n_rows);
    for(uword i = 0; i < outputs.n_rows; ++i){
        vec outputs_copy = outputs.row(i).t();
        params.col(i) = weighted_linear_regression(features, outputs_copy, weights.row(i), reg_fact, normalize, bias_entry());
    }
}

/*** LINEAR REGRESSION ***/
mat LinFunc::feature_fn(const mat &raw) const{
    return join_cols(raw, mat(1, raw.n_cols, fill::ones));

}

/*** QUADRATIC REGRESSION ***/

QuadFunc::QuadFunc(uword input_dim, double reg_fact, bool normalize) :
    RegressionFunc(reg_fact, normalize),
    input_dim(input_dim)
{
    num_quad_features = static_cast<uword>(std::round((input_dim + 1) * (0.5 * input_dim)));
    feature_dim = num_quad_features + input_dim + 1;
    idx = find(trimatl(mat(input_dim, input_dim, fill::ones)));

    quad_term = mat(input_dim, input_dim, fill::none);
    lin_term = vec(input_dim, fill::none);
}

void QuadFunc::extract_terms() {
    vec params = get_params();
    quad_term.zeros();
    quad_term(idx) = params.rows(0, num_quad_features-1);
    quad_term = - quad_term - quad_term.t();

    lin_term = params.rows(num_quad_features, feature_dim-2);

    const_term = params.at(feature_dim - 1);
}

void QuadFunc::fit(const mat &inputs, const vec &outputs) {
    RegressionFunc::fit(inputs, outputs);
    extract_terms();
}

void QuadFunc::fit(const mat &inputs, const vec &outputs, const rowvec &weights) {
    RegressionFunc::fit(inputs, outputs, weights);
    extract_terms();
}
void QuadFunc::fit_whitened(const mat& inputs, const vec& outputs, const vec& sample_mean, const mat& sample_covar_chol) {
    mat inv_sample_covar_chol = inv(sample_covar_chol);
    mat white_inputs = whiten_inputs(inputs, sample_mean, inv_sample_covar_chol);
    RegressionFunc::fit(white_inputs, outputs);
    extract_terms(sample_mean, inv_sample_covar_chol);
}

void QuadFunc::fit_whitened(const mat& inputs, const vec& outputs, const rowvec &weights, const vec& sample_mean,
                            const mat& sample_covar_chol) {
    mat inv_sample_covar_chol = inv(sample_covar_chol);
    mat white_inputs = whiten_inputs(inputs, sample_mean, inv_sample_covar_chol);
    RegressionFunc::fit(white_inputs, outputs, weights);
    extract_terms(sample_mean, inv_sample_covar_chol);
}

mat QuadFunc::whiten_inputs(const mat& inputs, const vec& sample_mean, const mat& inv_sample_covar_chol) {
    return inv_sample_covar_chol * (inputs.each_col() - sample_mean);
}

void  QuadFunc::extract_terms(const vec& sample_mean, const mat& inv_sample_covar_chol) {
    vec params = get_params();
    quad_term.zeros();
    quad_term(idx) = params.rows(0, num_quad_features-1);
    quad_term = - quad_term - quad_term.t();
    quad_term = inv_sample_covar_chol.t() * quad_term * inv_sample_covar_chol;

    lin_term = params.rows(num_quad_features, feature_dim-2);
    vec b = inv_sample_covar_chol.t() * lin_term;
    lin_term = quad_term * sample_mean + b;
    const_term = params.at(feature_dim - 1) + dot(sample_mean, - 0.5 * quad_term * sample_mean - b);
}

File: EIM/util/Regression.py
import numpy as np


class RegressionFunc:

    def __init__(self, reg_fact, normalize, unnormalize_output, bias_entry=None):
        self._reg_fact = reg_fact
        self._normalize = normalize
        self._unnormalize_output = unnormalize_output
        self._bias_entry = bias_entry
        self._params = None
        self.o_std = None

    def __call__(self, inputs):
        if self._params is None:
            raise AssertionError("Model not trained yet")
        return self._feature_fn(inputs) @ self._params

    def _feature_fn(self, x):
        raise NotImplementedError

    def _normalize_features(self, features):
        mean = np.mean(features, axis=0, keepdims=True)
        std = np.std(features, axis=0, keepdims=True)
        # do not normalize bias
        if self._bias_entry is not None:
            mean[self._bias_entry] = 0.0
            std[self._bias_entry] = 1.0
        features = (features - mean) / std
        return features, np.squeeze(mean, axis=0), np.squeeze(std, axis=0)

    def _normalize_outputs(self, outputs):
        mean = np.mean(outputs)
        std = np.std(outputs)
        outputs = (outputs - mean) / std
        return outputs, mean, std

    def _undo_normalization(self, params, f_mean, f_std, o_mean, o_std):
        if self._unnormalize_output:
            params *= (o_std / f_std)
            params[self._bias_entry] = params[self._bias_entry] - np.dot(params, f_mean) + o_mean
        else:
            params *= (1.0 / f_std)
            params[self._bias_entry] = params[self._bias_entry] - np.dot(params, f_mean)
        return params

    def fit(self, inputs, outputs, weights=None):
        if len(outputs.shape) > 1:
            outputs = np.squeeze(outputs)
        features = self._feature_fn(inputs)
        if self._normalize:
            features, f_mean, f_std = self._normalize_features(features)
            outputs, o_mean, o_std = self._normalize_outputs(outputs)

        if weights is not None:
            if len(weights.shape) == 1:
                weights = np.expand_dims(weights, 1)
            weighted_features = weights * features
        else:
            weighted_features = features

        #regression
        reg_mat = np.eye(weighted_features.shape[-1]) * self._reg_fact
        if self._bias_entry is not None:
            reg_mat[self._bias_entry, self._bias_entry] = 0.0
        try:
            self._params = np.linalg.solve(weighted_features.T @ features + reg_mat, weighted_features.T @ outputs)
            if self._normalize:
                self._undo_normalization(self._params, f_mean, f_std, o_mean, o_std)
                self.o_std = o_std
        except np.linalg.LinAlgError as e:
            print("Error during matrix inversion", e.what())


class LinFunc(RegressionFunc):

    def __init__(self, reg_fact, normalize, unnormalize_output):
        super().__init__(reg_fact, normalize, unnormalize_output, -1)

    def _feature_fn(self, x):
        return np.concatenate([x, np.ones([x.shape[0], 1], dtype=x.dtype)], 1)


class QuadFunc(RegressionFunc):
    #*Fits - 0.5 * x ^ T  Rx + x ^ T r + r_0 ** * /

    def __init__(self, reg_fact, normalize, unnormalize_output):
        super().__init__(reg_fact, normalize, unnormalize_output, bias_entry=-1)
        self.quad_term = None
        self.lin_term = None
        self.const_term = None
        self._square_feat_upper_tri_ind = None

    def __square_feat_upper_tri_ind(self, n):
        if self._square_feat_upper_tri_ind is None or self._square_feat_upper_tri_ind.shape[0] != n:
            self._square_feat_upper_tri_ind = np.triu_indices(n)
        return self._square_feat_upper_tri_ind

    def _feature_fn(self, x):
        lin_feat = x
        quad_feat = np.transpose((x[:, :, np.newaxis] @ x[:, np.newaxis, :]), [1, 2, 0])
        quad_feat = quad_feat[self.__square_feat_upper_tri_ind(x.shape[-1])].T
        features = np.hstack([quad_feat, lin_feat, np.ones([x.shape[0], 1])])
        return features

    def fit(self, inputs, outputs, weights=None, sample_mean=None, sample_chol_cov=None):
        if sample_mean is None:
            assert sample_chol_cov is None
        if sample_chol_cov is None:
            assert sample_mean is None

        #whithening
        if sample_mean is not None and sample_chol_cov is not None:
            inv_samples_chol_cov = np.linalg.inv(sample_chol_cov)
            inputs = (inputs - sample_mean) @ inv_samples_chol_cov.T

        dim = inputs.shape[-1]

        super().fit(inputs, outputs, weights)

        idx = np.triu(np.ones([dim, dim], np.bool))

        qt = np.zeros([dim, dim])
        qt[idx] = self._params[:- (dim + 1)]
        self.quad_term = - qt - qt.T

        self.lin_term = self._params[-(dim + 1): -1]
        self.const_term = self._params[-1]

        #unwhitening:
        if sample_mean is not None and sample_chol_cov is not None:
            self.quad_term = inv_samples_chol_cov.T @ self.quad_term @ inv_samples_chol_cov
            t1 = inv_samples_chol_cov.T @ self.lin_term
            t2 = self.quad_term @ sample_mean
            self.lin_term = t1 + t2
            self.const_term += np.dot(sample_mean, -0.5 * t2 - t1)
